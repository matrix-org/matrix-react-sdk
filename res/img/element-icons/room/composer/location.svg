<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M12 2C8.13 2 5 5.13 5 9C5 13.17 9.42 18.92 11.24 21.11C11.64 21.59 12.37 21.59 12.77 21.11C14.58 18.92 19 13.17 19 9C19 5.13 15.87 2 12 2ZM12 11.5C10.62 11.5 9.5 10.38 9.5 9C9.5 7.62 10.62 6.5 12 6.5C13.38 6.5 14.5 7.62 14.5 9C14.5 10.38 13.38 11.5 12 11.5Z" fill="#0DBD8B"/>
</svg>
