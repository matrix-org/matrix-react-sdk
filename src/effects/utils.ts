/*
 Copyright 2020 Nurjin Jafar
 Copyright 2020 Nordeck IT + Consulting GmbH.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
/**
 * Checks a message if it contains one of the provided emojis
 * @param  {Object} content The message
 * @param  {Array<string>} emojis The list of emojis to check for
 */
export const containsEmoji = (content: { msgtype: string, body: string }, emojis: Array<string>): boolean => {
    return emojis.some((emoji) => content.body && content.body.includes(emoji));
}
