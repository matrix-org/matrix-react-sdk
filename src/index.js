/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var Skinner = require('./Skinner');

var t = require('counterpart');

t.registerTranslations('en', require('../languages/en'));
t.registerTranslations('pt-BR', require('../languages/pt-BR'));

t.setFallbackLocale('en');
t.setLocale(window.navigator.languages[0]);



module.exports.loadSkin = function(skinObject) {
    Skinner.load(skinObject);
};

module.exports.resetSkin = function() {
    Skinner.reset();
};

module.exports.getComponent = function(componentName) {
    return Skinner.getComponent(componentName);
};
