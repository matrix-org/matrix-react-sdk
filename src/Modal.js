/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/


import React from 'react';
import ReactDOM from 'react-dom';
import Analytics from './Analytics';
import dis from './dispatcher/dispatcher';
import {defer} from './utils/promise';
import AsyncWrapper from './AsyncWrapper';

const DIALOG_CONTAINER_ID = "mx_Dialog_Container";
const STATIC_DIALOG_CONTAINER_ID = "mx_Dialog_StaticContainer";

class ModalManager {
    constructor() {
        this._counter = 0;

        // The modal to prioritise over all others. If this is set, only show
        // this modal. Remove all other modals from the stack when this modal
        // is closed.
        this._priorityModal = null;
        // The modal to keep open underneath other modals if possible. Useful
        // for cases like Settings where the modal should remain open while the
        // user is prompted for more information/errors.
        this._staticModal = null;
        // A list of the modals we have stacked up, with the most recent at [0]
        // Neither the static nor priority modal will be in this list.
        this._modals = [
            /* {
               elem: React component for this dialog
               onFinished: caller-supplied onFinished callback
               className: CSS class for the dialog wrapper div
               } */
        ];

        this.onBackgroundClick = this.onBackgroundClick.bind(this);
    }

    hasDialogs() {
        return this._priorityModal || this._staticModal || this._modals.length > 0;
    }

    getOrCreateContainer() {
        let container = document.getElementById(DIALOG_CONTAINER_ID);

        if (!container) {
            container = document.createElement("div");
            container.id = DIALOG_CONTAINER_ID;
            document.body.appendChild(container);
        }

        return container;
    }

    getOrCreateStaticContainer() {
        let container = document.getElementById(STATIC_DIALOG_CONTAINER_ID);

        if (!container) {
            container = document.createElement("div");
            container.id = STATIC_DIALOG_CONTAINER_ID;
            document.body.appendChild(container);
        }

        return container;
    }

    createTrackedDialog(analyticsAction, analyticsInfo, ...rest) {
        Analytics.trackEvent('Modal', analyticsAction, analyticsInfo);
        return this.createDialog(...rest);
    }

    appendTrackedDialog(analyticsAction, analyticsInfo, ...rest) {
        Analytics.trackEvent('Modal', analyticsAction, analyticsInfo);
        return this.appendDialog(...rest);
    }

    createDialog(Element, ...rest) {
        return this.createDialogAsync(Promise.resolve(Element), ...rest);
    }

    appendDialog(Element, ...rest) {
        return this.appendDialogAsync(Promise.resolve(Element), ...rest);
    }

    createTrackedDialogAsync(analyticsAction, analyticsInfo, ...rest) {
        Analytics.trackEvent('Modal', analyticsAction, analyticsInfo);
        return this.createDialogAsync(...rest);
    }

    appendTrackedDialogAsync(analyticsAction, analyticsInfo, ...rest) {
        Analytics.trackEvent('Modal', analyticsAction, analyticsInfo);
        return this.appendDialogAsync(...rest);
    }

    _buildModal(prom, props, className, options) {
        const modal = {};

        // never call this from onFinished() otherwise it will loop
        const [closeDialog, onFinishedProm] = this._getCloseFn(modal, props);

        // don't attempt to reuse the same AsyncWrapper for different dialogs,
        // otherwise we'll get confused.
        const modalCount = this._counter++;

        // FIXME: If a dialog uses getDefaultProps it clobbers the onFinished
        // property set here so you can't close the dialog from a button click!
        modal.elem = (
            <AsyncWrapper key={modalCount} prom={prom} {...props}
                          onFinished={closeDialog} />
        );
        modal.onFinished = props ? props.onFinished : null;
        modal.className = className;
        modal.onBeforeClose = options.onBeforeClose;
        modal.beforeClosePromise = null;
        modal.close = closeDialog;
        modal.closeReason = null;

        return {modal, closeDialog, onFinishedProm};
    }

    _getCloseFn(modal, props) {
        const deferred = defer();
        return [async (...args) => {
            if (modal.beforeClosePromise) {
                await modal.beforeClosePromise;
            } else if (modal.onBeforeClose) {
                modal.beforeClosePromise = modal.onBeforeClose(modal.closeReason);
                const shouldClose = await modal.beforeClosePromise;
                modal.beforeClosePromise = null;
                if (!shouldClose) {
                    return;
                }
            }
            deferred.resolve(args);
            if (props && props.onFinished) props.onFinished.apply(null, args);
            const i = this._modals.indexOf(modal);
            if (i >= 0) {
                this._modals.splice(i, 1);
            }

            if (this._priorityModal === modal) {
                this._priorityModal = null;

                // XXX: This is destructive
                this._modals = [];
            }

            if (this._staticModal === modal) {
                this._staticModal = null;

                // XXX: This is destructive
                this._modals = [];
            }

            this._reRender();
        }, deferred.promise];
    }

    /**
     * @callback onBeforeClose
     * @param {string?} reason either "backgroundClick" or null
     * @return {Promise<bool>} whether the dialog should close
     */

    /**
     * Open a modal view.
     *
     * This can be used to display a react component which is loaded as an asynchronous
     * webpack component. To do this, set 'loader' as:
     *
     *   (cb) => {
     *       require(['<module>'], cb);
     *   }
     *
     * @param {Promise} prom   a promise which resolves with a React component
     *   which will be displayed as the modal view.
     *
     * @param {Object} props   properties to pass to the displayed
     *    component. (We will also pass an 'onFinished' property.)
     *
     * @param {String} className   CSS class to apply to the modal wrapper
     *
     * @param {boolean} isPriorityModal if true, this modal will be displayed regardless
     *                                  of other modals that are currently in the stack.
     *                                  Also, when closed, all modals will be removed
     *                                  from the stack.
     * @param {boolean} isStaticModal  if true, this modal will be displayed under other
     *                                 modals in the stack. When closed, all modals will
     *                                 also be removed from the stack. This is not compatible
     *                                 with being a priority modal. Only one modal can be
     *                                 static at a time.
     * @param {Object} options? extra options for the dialog
     * @param {onBeforeClose} options.onBeforeClose a callback to decide whether to close the dialog
     * @returns {object} Object with 'close' parameter being a function that will close the dialog
     */
    createDialogAsync(prom, props, className, isPriorityModal, isStaticModal, options = {}) {
        const {modal, closeDialog, onFinishedProm} = this._buildModal(prom, props, className, options);
        if (isPriorityModal) {
            // XXX: This is destructive
            this._priorityModal = modal;
        } else if (isStaticModal) {
            // This is intentionally destructive
            this._staticModal = modal;
        } else {
            this._modals.unshift(modal);
        }

        this._reRender();
        return {
            close: closeDialog,
            finished: onFinishedProm,
        };
    }

    appendDialogAsync(prom, props, className) {
        const {modal, closeDialog, onFinishedProm} = this._buildModal(prom, props, className, {});

        this._modals.push(modal);
        this._reRender();
        return {
            close: closeDialog,
            finished: onFinishedProm,
        };
    }

    onBackgroundClick() {
        const modal = this._getCurrentModal();
        if (!modal) {
            return;
        }
        // we want to pass a reason to the onBeforeClose
        // callback, but close is currently defined to
        // pass all number of arguments to the onFinished callback
        // so, pass the reason to close through a member variable
        modal.closeReason = "backgroundClick";
        modal.close();
        modal.closeReason = null;
    }

    _getCurrentModal() {
        return this._priorityModal ? this._priorityModal : (this._modals[0] || this._staticModal);
    }

    _reRender() {
        if (this._modals.length === 0 && !this._priorityModal && !this._staticModal) {
            // If there is no modal to render, make all of Riot available
            // to screen reader users again
            dis.dispatch({
                action: 'aria_unhide_main_app',
            });
            ReactDOM.unmountComponentAtNode(this.getOrCreateContainer());
            ReactDOM.unmountComponentAtNode(this.getOrCreateStaticContainer());
            return;
        }

        // Hide the content outside the modal to screen reader users
        // so they won't be able to navigate into it and act on it using
        // screen reader specific features
        dis.dispatch({
            action: 'aria_hide_main_app',
        });

        if (this._staticModal) {
            const classes = "mx_Dialog_wrapper mx_Dialog_staticWrapper "
                + (this._staticModal.className ? this._staticModal.className : '');

            const staticDialog = (
                <div className={classes}>
                    <div className="mx_Dialog">
                        { this._staticModal.elem }
                    </div>
                    <div className="mx_Dialog_background mx_Dialog_staticBackground" onClick={this.onBackgroundClick}></div>
                </div>
            );

            ReactDOM.render(staticDialog, this.getOrCreateStaticContainer());
        } else {
            // This is safe to call repeatedly if we happen to do that
            ReactDOM.unmountComponentAtNode(this.getOrCreateStaticContainer());
        }

        const modal = this._getCurrentModal();
        if (modal !== this._staticModal) {
            const classes = "mx_Dialog_wrapper "
                + (this._staticModal ? "mx_Dialog_wrapperWithStaticUnder " : '')
                + (modal.className ? modal.className : '');

            const dialog = (
                <div className={classes}>
                    <div className="mx_Dialog">
                        {modal.elem}
                    </div>
                    <div className="mx_Dialog_background" onClick={this.onBackgroundClick}></div>
                </div>
            );

            ReactDOM.render(dialog, this.getOrCreateContainer());
        } else {
            // This is safe to call repeatedly if we happen to do that
            ReactDOM.unmountComponentAtNode(this.getOrCreateContainer());
        }
    }
}

if (!global.singletonModalManager) {
    global.singletonModalManager = new ModalManager();
}
export default global.singletonModalManager;
