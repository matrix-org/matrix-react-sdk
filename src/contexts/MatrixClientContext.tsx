/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, {
    ComponentClass,
    createContext,
    forwardRef,
    PropsWithoutRef,
    ForwardRefExoticComponent,
    useContext,
    RefAttributes,
} from "react";
import { MatrixClient } from "matrix-js-sdk/src/matrix";

// This context is available to components under LoggedInView,
// the context must not be used by components outside a MatrixClientContext tree.
// This assertion allows us to make the type not nullable.
const MatrixClientContext = createContext<MatrixClient>(null as any);
MatrixClientContext.displayName = "MatrixClientContext";
export default MatrixClientContext;

export interface MatrixClientProps {
    mxClient: MatrixClient;
}

export function useMatrixClientContext(): MatrixClient {
    return useContext(MatrixClientContext);
}

const matrixHOC = <ComposedComponentProps extends {}>(
    ComposedComponent: ComponentClass<ComposedComponentProps>,
): ForwardRefExoticComponent<
    PropsWithoutRef<Omit<ComposedComponentProps, "mxClient">> &
        RefAttributes<InstanceType<ComponentClass<ComposedComponentProps>>>
> => {
    type ComposedComponentInstance = InstanceType<typeof ComposedComponent>;

    // eslint-disable-next-line react-hooks/rules-of-hooks

    const TypedComponent = ComposedComponent;

    return forwardRef<ComposedComponentInstance, Omit<ComposedComponentProps, "mxClient">>((props, ref) => {
        const client = useContext(MatrixClientContext);

        // @ts-ignore
        return <TypedComponent ref={ref} {...props} mxClient={client} />;
    });
};
export const withMatrixClientHOC = matrixHOC;
