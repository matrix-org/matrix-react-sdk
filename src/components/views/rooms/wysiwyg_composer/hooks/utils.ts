/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { MutableRefObject } from "react";

import { TimelineRenderingType } from "../../../../../contexts/RoomContext";
import { IRoomState } from "../../../../structures/RoomView";

export function focusComposer(
    composerElement: MutableRefObject<HTMLElement | null>,
    renderingType: TimelineRenderingType,
    roomContext: IRoomState,
    timeoutId: MutableRefObject<number | null>,
): void {
    if (renderingType === roomContext.timelineRenderingType) {
        // Immediately set the focus, so if you start typing it
        // will appear in the composer
        composerElement.current?.focus();
        // If we call focus immediate, the focus _is_ in the right
        // place, but the cursor is invisible, presumably because
        // some other event is still processing.
        // The following line ensures that the cursor is actually
        // visible in composer.
        if (timeoutId.current) {
            clearTimeout(timeoutId.current);
        }
        timeoutId.current = window.setTimeout(() => composerElement.current?.focus(), 200);
    }
}

export function setCursorPositionAtTheEnd(element: HTMLElement): void {
    const range = document.createRange();
    range.selectNodeContents(element);
    range.collapse(false);
    const selection = document.getSelection();
    selection.removeAllRanges();
    selection.addRange(range);

    element.focus();
}
