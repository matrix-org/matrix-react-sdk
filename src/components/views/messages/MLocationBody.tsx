/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import maplibregl from 'maplibre-gl';
import { logger } from "matrix-js-sdk/src/logger";
import { LOCATION_EVENT_TYPE } from 'matrix-js-sdk/src/@types/location';
import { MatrixEvent } from 'matrix-js-sdk/src/models/event';

import SdkConfig from '../../../SdkConfig';
import { replaceableComponent } from "../../../utils/replaceableComponent";
import { IBodyProps } from "./IBodyProps";
import { _t } from '../../../languageHandler';
import MemberAvatar from '../avatars/MemberAvatar';

interface IState {
    error: Error;
}

@replaceableComponent("views.messages.MLocationBody")
export default class MLocationBody extends React.Component<IBodyProps, IState> {
    private map: maplibregl.Map;
    private coords: GeolocationCoordinates;

    constructor(props: IBodyProps) {
        super(props);

        // unfortunately we're stuck supporting legacy `content.geo_uri`
        // events until the end of days, or until we figure out mutable
        // events - so folks can read their old chat history correctly.
        // https://github.com/matrix-org/matrix-doc/issues/3516
        const content = this.props.mxEvent.getContent();
        const loc = content[LOCATION_EVENT_TYPE.name];
        const uri = loc ? loc.uri : content['geo_uri'];

        this.coords = parseGeoUri(uri);
        this.state = {
            error: undefined,
        };
    }

    componentDidMount() {
        const config = SdkConfig.get();
        const coordinates = new maplibregl.LngLat(
            this.coords.longitude, this.coords.latitude);

        this.map = new maplibregl.Map({
            container: this.getBodyId(),
            style: config.map_style_url,
            center: coordinates,
            zoom: 13,
        });

        new maplibregl.Marker({
            element: document.getElementById(this.getMarkerId()),
            anchor: 'bottom',
            offset: [0, -1],
        })
            .setLngLat(coordinates)
            .addTo(this.map);

        this.map.on('error', (e)=>{
            logger.error(
                "Failed to load map: check map_style_url in config.json has a "
                + "valid URL and API key",
                e.error,
            );
            this.setState({ error: e.error });
        });
    }

    private getBodyId = () => {
        return `mx_MLocationBody_${this.props.mxEvent.getId()}`;
    };

    private getMarkerId = () => {
        return `mx_MLocationBody_marker_${this.props.mxEvent.getId()}`;
    };

    render() {
        const error = this.state.error ?
            <div className="mx_EventTile_tileError mx_EventTile_body">
                { _t("Failed to load map") }
            </div> : null;

        return <div className="mx_MLocationBody">
            <div id={this.getBodyId()} className="mx_MLocationBody_map" />
            { error }
            <div className="mx_MLocationBody_marker" id={this.getMarkerId()}>
                <div className="mx_MLocationBody_markerBorder">
                    <MemberAvatar
                        member={this.props.mxEvent.sender}
                        width={27}
                        height={27}
                        viewUserOnClick={false}
                    />
                </div>
                <img
                    className="mx_MLocationBody_pointer"
                    src={require("../../../../res/img/location/pointer.svg")}
                    width="9"
                    height="5"
                />
            </div>
        </div>;
    }
}

export function parseGeoUri(uri: string): GeolocationCoordinates {
    function parse(s: string): number {
        const ret = parseFloat(s);
        if (Number.isNaN(ret)) {
            return undefined;
        } else {
            return ret;
        }
    }

    const m = uri.match(/^\s*geo:(.*?)\s*$/);
    if (!m) return;
    const parts = m[1].split(';');
    const coords = parts[0].split(',');
    let uncertainty: number;
    for (const param of parts.slice(1)) {
        const m = param.match(/u=(.*)/);
        if (m) uncertainty = parse(m[1]);
    }
    return {
        latitude: parse(coords[0]),
        longitude: parse(coords[1]),
        altitude: parse(coords[2]),
        accuracy: uncertainty,
        altitudeAccuracy: undefined,
        heading: undefined,
        speed: undefined,
    };
}

function makeLink(coords: GeolocationCoordinates): string {
    return (
        "https://www.openstreetmap.org/" +
        `?mlat=${coords.latitude}` +
        `&mlon=${coords.longitude}` +
        `#map=16/${coords.latitude}/${coords.longitude}`
    );
}

export function createMapSiteLink(event: MatrixEvent): string {
    const content: Object = event.getContent();
    const mLocation = content[LOCATION_EVENT_TYPE.name];
    if (mLocation !== undefined) {
        const uri = mLocation["uri"];
        if (uri !== undefined) {
            return makeLink(parseGeoUri(uri));
        }
    } else {
        const geoUri = content["geo_uri"];
        if (geoUri) {
            return makeLink(parseGeoUri(geoUri));
        }
    }
    return null;
}
