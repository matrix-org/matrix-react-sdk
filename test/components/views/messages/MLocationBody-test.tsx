/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
// eslint-disable-next-line deprecate/import
import { mount } from "enzyme";
import { LocationAssetType } from "matrix-js-sdk/src/@types/location";
import { ClientEvent, RoomMember } from 'matrix-js-sdk/src/matrix';
import maplibregl from 'maplibre-gl';
import { logger } from 'matrix-js-sdk/src/logger';
import { act } from 'react-dom/test-utils';
import { SyncState } from 'matrix-js-sdk/src/sync';

import MLocationBody from "../../../../src/components/views/messages/MLocationBody";
import MatrixClientContext from "../../../../src/contexts/MatrixClientContext";
import { RoomPermalinkCreator } from "../../../../src/utils/permalinks/Permalinks";
import { MediaEventHelper } from "../../../../src/utils/MediaEventHelper";
import Modal from '../../../../src/Modal';
import SdkConfig from "../../../../src/SdkConfig";
import { TILE_SERVER_WK_KEY } from '../../../../src/utils/WellKnownUtils';
import { makeLocationEvent } from "../../../test-utils/location";
import { getMockClientWithEventEmitter } from '../../../test-utils';

describe("MLocationBody", () => {
    describe('<MLocationBody>', () => {
        const roomId = '!room:server';
        const userId = '@user:server';
        const mockClient = getMockClientWithEventEmitter({
            getClientWellKnown: jest.fn().mockReturnValue({
                [TILE_SERVER_WK_KEY.name]: { map_style_url: 'maps.com' },
            }),
            isGuest: jest.fn().mockReturnValue(false),
        });
        const defaultEvent = makeLocationEvent("geo:51.5076,-0.1276", LocationAssetType.Pin);
        const defaultProps = {
            mxEvent: defaultEvent,
            highlights: [],
            highlightLink: '',
            onHeightChanged: jest.fn(),
            onMessageAllowed: jest.fn(),
            permalinkCreator: {} as RoomPermalinkCreator,
            mediaEventHelper: {} as MediaEventHelper,
        };
        const getComponent = (props = {}) => mount(<MLocationBody {...defaultProps} {...props} />, {
            wrappingComponent: MatrixClientContext.Provider,
            wrappingComponentProps: { value: mockClient },
        });
        const getMapErrorComponent = () => {
            const mockMap = new maplibregl.Map();
            mockClient.getClientWellKnown.mockReturnValue({
                [TILE_SERVER_WK_KEY.name]: { map_style_url: 'bad-tile-server.com' },
            });
            const component = getComponent();

            // simulate error initialising map in maplibregl
            // @ts-ignore
            mockMap.emit('error', { status: 404 });

            return component;
        };

        beforeAll(() => {
            maplibregl.AttributionControl = jest.fn();
        });

        beforeEach(() => {
            jest.clearAllMocks();
        });

        describe('with error', () => {
            let sdkConfigSpy;

            beforeEach(() => {
                // eat expected errors to keep console clean
                jest.spyOn(logger, 'error').mockImplementation(() => { });
                mockClient.getClientWellKnown.mockReturnValue({});
                sdkConfigSpy = jest.spyOn(SdkConfig, 'get').mockReturnValue({});
            });

            afterAll(() => {
                sdkConfigSpy.mockRestore();
                jest.spyOn(logger, 'error').mockRestore();
            });

            it('displays correct fallback content without error style when map_style_url is not configured', () => {
                const component = getComponent();
                expect(component.find(".mx_EventTile_body")).toMatchSnapshot();
            });

            it('displays correct fallback content when map_style_url is misconfigured', () => {
                const component = getMapErrorComponent();
                component.setProps({});
                expect(component.find(".mx_EventTile_body")).toMatchSnapshot();
            });

            it('should clear the error on reconnect', () => {
                const component = getMapErrorComponent();
                expect((component.state() as React.ComponentState).error).toBeDefined();
                mockClient.emit(ClientEvent.Sync, SyncState.Reconnecting, SyncState.Error);
                expect((component.state() as React.ComponentState).error).toBeUndefined();
            });
        });

        describe('without error', () => {
            beforeEach(() => {
                mockClient.getClientWellKnown.mockReturnValue({
                    [TILE_SERVER_WK_KEY.name]: { map_style_url: 'maps.com' },
                });

                // MLocationBody uses random number for map id
                // stabilise for test
                jest.spyOn(global.Math, 'random').mockReturnValue(0.123456);
            });

            afterAll(() => {
                jest.spyOn(global.Math, 'random').mockRestore();
            });

            it('renders map correctly', () => {
                const mockMap = new maplibregl.Map();
                const component = getComponent();

                expect(component).toMatchSnapshot();
                // map was centered
                expect(mockMap.setCenter).toHaveBeenCalledWith({
                    lat: 51.5076, lon: -0.1276,
                });
            });

            it('opens map dialog on click', () => {
                const modalSpy = jest.spyOn(Modal, 'createDialog').mockReturnValue(undefined);
                const component = getComponent();

                act(() => {
                    component.find('Map').at(0).simulate('click');
                });

                expect(modalSpy).toHaveBeenCalled();
            });

            it('renders marker correctly for a non-self share', () => {
                const mockMap = new maplibregl.Map();
                const component = getComponent();

                expect(component.find('SmartMarker').at(0).props()).toEqual(
                    expect.objectContaining({
                        map: mockMap,
                        geoUri: 'geo:51.5076,-0.1276',
                        roomMember: undefined,
                    }),
                );
            });

            it('renders marker correctly for a self share', () => {
                const selfShareEvent = makeLocationEvent("geo:51.5076,-0.1276", LocationAssetType.Self);
                const member = new RoomMember(roomId, userId);
                // @ts-ignore cheat assignment to property
                selfShareEvent.sender = member;
                const component = getComponent({ mxEvent: selfShareEvent });

                // render self locations with user avatars
                expect(component.find('SmartMarker').at(0).props()['roomMember']).toEqual(
                    member,
                );
            });
        });
    });
});
