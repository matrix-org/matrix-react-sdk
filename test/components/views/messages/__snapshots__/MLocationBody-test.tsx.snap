// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MLocationBody <MLocationBody> with error displays correct fallback content when map_style_url is misconfigured 1`] = `
<div
  className="mx_EventTile_body mx_MLocationBody"
>
  <span
    className="mx_EventTile_tileError"
  >
    Unable to load map: This homeserver is not configured correctly to display maps, or the configured map server may be unreachable.
  </span>
  <br />
  Shared a location: Found at geo:51.5076,-0.1276 at 2021-12-21T12:22+0000
</div>
`;

exports[`MLocationBody <MLocationBody> with error displays correct fallback content without error style when map_style_url is not configured 1`] = `
<div
  className="mx_EventTile_body mx_MLocationBody"
>
  <span
    className=""
  >
    Unable to load map: This homeserver is not configured to display maps.
  </span>
  <br />
  Shared a location: Found at geo:51.5076,-0.1276 at 2021-12-21T12:22+0000
</div>
`;

exports[`MLocationBody <MLocationBody> without error renders map correctly 1`] = `
<MLocationBody
  highlightLink=""
  highlights={Array []}
  mediaEventHelper={Object {}}
  mxEvent={
    Object {
      "content": Object {
        "body": "Found at geo:51.5076,-0.1276 at 2021-12-21T12:22+0000",
        "geo_uri": "geo:51.5076,-0.1276",
        "msgtype": "m.location",
        "org.matrix.msc1767.text": "Found at geo:51.5076,-0.1276 at 2021-12-21T12:22+0000",
        "org.matrix.msc3488.asset": Object {
          "type": "m.pin",
        },
        "org.matrix.msc3488.location": Object {
          "description": "Human-readable label",
          "uri": "geo:51.5076,-0.1276",
        },
        "org.matrix.msc3488.ts": 252523,
      },
      "event_id": "$2",
      "type": "org.matrix.msc3488.location",
    }
  }
  onHeightChanged={[MockFunction]}
  onMessageAllowed={[MockFunction]}
  permalinkCreator={Object {}}
>
  <LocationBodyContent
    mapId="mx_MLocationBody_$2_HHHHHHHH"
    mxEvent={
      Object {
        "content": Object {
          "body": "Found at geo:51.5076,-0.1276 at 2021-12-21T12:22+0000",
          "geo_uri": "geo:51.5076,-0.1276",
          "msgtype": "m.location",
          "org.matrix.msc1767.text": "Found at geo:51.5076,-0.1276 at 2021-12-21T12:22+0000",
          "org.matrix.msc3488.asset": Object {
            "type": "m.pin",
          },
          "org.matrix.msc3488.location": Object {
            "description": "Human-readable label",
            "uri": "geo:51.5076,-0.1276",
          },
          "org.matrix.msc3488.ts": 252523,
        },
        "event_id": "$2",
        "type": "org.matrix.msc3488.location",
      }
    }
    onClick={[Function]}
    onError={[Function]}
    tooltip="Expand map"
  >
    <div
      className="mx_MLocationBody"
    >
      <TooltipTarget
        alignment={5}
        label="Expand map"
        maxParentWidth={450}
      >
        <div
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          onMouseOver={[Function]}
          tabIndex={0}
        >
          <Map
            centerGeoUri="geo:51.5076,-0.1276"
            className="mx_MLocationBody_map"
            id="mx_MLocationBody_$2_HHHHHHHH"
            onClick={[Function]}
            onError={[Function]}
          >
            <div
              className="mx_Map mx_MLocationBody_map"
              id="mx_Map_mx_MLocationBody_$2_HHHHHHHH"
              onClick={[Function]}
            >
              <SmartMarker
                geoUri="geo:51.5076,-0.1276"
                id="mx_MLocationBody_$2_HHHHHHHH-marker"
                map={
                  MockMap {
                    "_events": Object {
                      "error": Array [
                        [Function],
                        [Function],
                        [Function],
                        [Function],
                        [Function],
                        [Function],
                      ],
                    },
                    "_eventsCount": 1,
                    "_maxListeners": undefined,
                    "addControl": [MockFunction] {
                      "calls": Array [
                        Array [
                          mockConstructor {},
                          "top-right",
                        ],
                        Array [
                          mockConstructor {},
                          "top-right",
                        ],
                      ],
                      "results": Array [
                        Object {
                          "type": "return",
                          "value": undefined,
                        },
                        Object {
                          "type": "return",
                          "value": undefined,
                        },
                      ],
                    },
                    "fitBounds": [MockFunction],
                    "removeControl": [MockFunction],
                    "setCenter": [MockFunction] {
                      "calls": Array [
                        Array [
                          Object {
                            "lat": 51.5076,
                            "lon": -0.1276,
                          },
                        ],
                        Array [
                          Object {
                            "lat": 51.5076,
                            "lon": -0.1276,
                          },
                        ],
                      ],
                      "results": Array [
                        Object {
                          "type": "return",
                          "value": undefined,
                        },
                        Object {
                          "type": "return",
                          "value": undefined,
                        },
                      ],
                    },
                    "setStyle": [MockFunction],
                    "zoomIn": [MockFunction],
                    "zoomOut": [MockFunction],
                    Symbol(kCapture): false,
                  }
                }
              >
                <span>
                  <ForwardRef
                    id="mx_MLocationBody_$2_HHHHHHHH-marker"
                  >
                    <div
                      className="mx_Marker mx_Marker_defaultColor"
                      id="mx_MLocationBody_$2_HHHHHHHH-marker"
                    >
                      <OptionalTooltip>
                        <div
                          className="mx_Marker_border"
                        >
                          <div
                            className="mx_Marker_icon"
                          />
                        </div>
                      </OptionalTooltip>
                    </div>
                  </ForwardRef>
                </span>
              </SmartMarker>
            </div>
          </Map>
        </div>
      </TooltipTarget>
    </div>
  </LocationBodyContent>
</MLocationBody>
`;
